src/marshmallow_annotations/builder.py

```diff
@@ -0,0 +1,236 @@
+
+
+class MarshmallowAnnotationError(MarshmallowError):
+    pass
+
+
+class AnnotationConversionError(MarshmallowAnnotationError):
+    pass
+
+
+def default_field_constructor(field):
+
+    def _(converter, subtypes, k):
+        return field(**k)
+
+    _.__name__ = f"{field.__name__}FieldConstructor"
+    return _
+
+
+def default_scheme_constructor(scheme_name):
+
+    def _(converter, subtypes, k):
+        return fields.Nested(scheme_name, **k)
+
+    _.__name__ = f"{scheme_name}FieldConstructor"
+    return _
+
+
+def _list_converter(converter, subtypes, k):
+    return fields.List(converter(subtypes[0]), **k)
+
+
+class TypeRegistry:
+    _registry = {
+        k: default_field_constructor(v)
+        for k, v in {
+            bool: fields.Boolean,
+            date: fields.Date,
+            datetime: fields.DateTime,
+            Decimal: fields.Decimal,
+            float: fields.Float,
+            int: fields.Integer,
+            str: fields.String,
+            time: fields.Time,
+            timedelta: fields.TimeDelta,
+            UUID: fields.UUID
+        }.items()
+    }
+
+    _registry[List] = _list_converter
+
+    registry = MappingProxyType(_registry)
+
+    @classmethod
+    def register(cls, type, field_constructor):
+        cls._registry[type] = field_constructor
+
+    @classmethod
+    def field_constructor(cls, type):
+
+        def field_constructor(constructor):
+            cls.register(type, constructor)
+            return constructor
+
+        return field_constructor
+
+    @classmethod
+    def from_type(cls, type):
+        converter = cls._registry.get(type)
+        if converter is None:
+            raise AnnotationConversionError(
+                f'No field factory found for {type!r}'
+            )
+        return converter
+
+    @classmethod
+    def register_field_for_type(cls, type, field):
+        cls.register(type, default_field_constructor(field))
+
+    @classmethod
+    def register_scheme_constructor(cls, type, scheme_or_name):
+        cls.register(type, default_scheme_constructor(scheme_or_name))
+
+
+
+
+def _is_optional(typehint):
+    # only supports single type optionals/unions
+    # as for the implementation... look, don't ask me
+    return _UNION_TYPE in type(typehint).__mro__ and len(
+        typehint.__args__
+    ) == 2 and typehint.__args__[1] is NoneType
+
+
+class BaseConverter:
+
+    def __init__(self, scheme, *, registry=TypeRegistry):
+        self.registry = TypeRegistry
+        self.scheme = scheme
+
+    def _field_from_typehint(self, typehint, kwargs=None):
+        # need that immutable dict in the stdlib pls
+        kwargs = kwargs if kwargs is not None else {}
+
+        # sane defaults
+        allow_none = False
+        required = True
+
+        if _is_optional(typehint):
+            allow_none = True
+            typehint = typehint.__args__[0]
+
+        # set this after optional check
+        subtypes = getattr(typehint, '__args__', ())
```

Comment:
What happens if this is a typevar, optional or union? Or do I punt that to whoever is doing that?